src/mds/MDCache.cc

```diff
@@ -8358,6 +8358,40 @@ int MDCache::path_traverse(const MDRequestRef& mdr, MDSContextFactory& cf,
   if (flags & MDS_TRAVERSE_CHECK_LOCKCACHE)
     mds->locker->find_and_attach_lock_cache(mdr, cur);
 
+  if (mdr) {
+    const cref_t<MClientRequest> &req = mdr->client_request;
+    Session *session = mds->get_session(req);
+    if (session && !session->info.has_feature(CEPHFS_FEATURE_ALTERNATE_NAME) &&
+	req->get_op() != CEPH_MDS_OP_LOOKUP &&
+	req->get_op() != CEPH_MDS_OP_LOOKUPHASH &&
+	req->get_op() != CEPH_MDS_OP_LOOKUPPARENT &&
+	req->get_op() != CEPH_MDS_OP_LOOKUPINO &&
+	req->get_op() != CEPH_MDS_OP_LOOKUPNAME &&
+	req->get_op() != CEPH_MDS_OP_LOOKUPSNAP &&
+	req->get_op() != CEPH_MDS_OP_RMSNAP &&
+	req->get_op() != CEPH_MDS_OP_LSSNAP &&
+	req->get_op() != CEPH_MDS_OP_GETATTR &&
+	req->get_op() != CEPH_MDS_OP_READDIR &&
+	req->get_op() != CEPH_MDS_OP_UNLINK &&
+	req->get_op() != CEPH_MDS_OP_RMDIR) {
+      MutationImpl::LockOpVec lov;
+      /* We need 'As' caps for the fscrypt context */
+      lov.add_rdlock(&in->authlock);
+      if (!mds->locker->acquire_locks(mdr, lov)) {
+	dout(10) << "traverse: failed to rdlock" << dendl;
+	return 1; /* XXX */
+      }
+      if (!in->get_inode()->fscrypt_auth.empty()) {
+	dout(10) << "blocking '" << ceph_mds_op_name(req->get_op())
+	  << "' operation in encrypted node" << dendl;
+	return -CEPHFS_EROFS;
+      }
+      /* Once we have made the fscrypt clients to flush the dirty caps
+       * back, we no longer need the authlock.
+       */
+      mds->locker->drop_lock(mdr.get(), &in->authlock);
```

Comment:
@lxbsz I guess this hunk is the change from the last push, yes? If the `As` cap wasn't dropped, that blocked reassigning caps to the requesting client?

Reply:
Yeah, correct.